docs/specifications/tests/DNSSEC-TP/dnssec16.md

```diff
@@ -0,0 +1,220 @@
+    [Method4] and [Method5] ("NS IP").
+
+4.  Create the following empty sets:
+    1.  Name server IP address and associated CDS RRset and its RRSIG
+        records ("CDS RRsets"). RRSIG records may be absent.
+    2.  Name server IP address and associated DNSKEY RRset and its 
+        RRsig records ("DNSKEY RRsets"). RRSIG records may be absent.
+    3.  Name server IP address ("No DNSKEY RRset").
+    4.  Name server IP address ("Mixed Delete CDS").
+    5.  Name server IP address ("Delete CDS").
+    6.  Name server IP address and associated CDS key tag 
+        ("No Match CDS With DNSKEY").
+    7.  Name server IP address and associated CDS key tag
+        ("DNSKEY Not Signed By CDS").
+    8.  Name server IP address ("CDS Not Signed").
+    9.  Name server IP address and key tag
+        ("CDS Signed By Unknown DNSKEY").
+    10. Name server IP address and key tag ("CDS Invalid RRSIG").
+
+5.  Repeat the following steps for each name server IP address in 
+    *NS IP*:
+
+    1. Send the CDS query over UDP to the name server IP address.
+       1. If no DNS response is returned, then go to next name server
+          IP.
+       2. Else, if AA bit is not set in the DNS response, then go to
+          next name server IP.
+       3. Else, if the RCODE in the DNS response is not *NOERROR*, then
+          go to next name server IP.
+       4. Else, if the DNS response contains at least one CDS record
+          in the answer section, then add the name server IP and the
+          CDS RRset to the *CDS RRsets* set. Also include any associated
+          RRSIG records.
+       5. Else, if the answer section has no CDS records, go to next
+          name server IP.
+    2. Send the DNSKEY query over UDP to the name server IP address.
+       1. If no DNS response is returned, then go to next name server
+          IP.
+       2. Else, if AA bit is not set in the DNS response, then go to
+          next name server IP.
+       3. Else, if the RCODE in the DNS response is not *NOERROR*, then
+          go to next name server IP.
+       4. Else, if the DNS response contains at least one DNSKEY
+          record in the answer section, then add the name server IP and
+          the DNSKEY RRset from the answer section to the 
+          *DNSKEY RRsets* set. Also include any associated RRSIG records.
+    3. Go to next name server IP.
+
+6.  If the *CDS RRsets* set is empty then terminate this test case.
+
+7.  For each name server IP in the *CDS RRsets* set do:
+
+    1. If the CDS RRset is non-empty, get the RRset and its RRSIG
+       records, else go to next name server IP address.
+    2. If any CDS record is a "delete" CDS, then do:
+       1. If there is more than a single CDS record then add the name
+          server IP to the *Mixed Delete CDS* set.
+       2. Else, add the name server IP address to the *Delete CDS*
+          set.
+       3. Go to next name server IP.
+    3. Get the DNSKEY and its RRSIG records, if any, from the
```

Comment:
What's the meaning of this "if any" clause again? Does it mean any DNSKEY and RRSIG records, or any RRSIG records for this DNSKEY since that set may be absent? Could it be present but empty? Anyway, what happens if the "if any" clause is false?

Reply:
Okay, so "if any" isn't about control flow. It's only there to let the reader know that they shouldn't worry even if there is no RRSIG RRset associated with the name server IP. Right?

How do you feel about this wording?

```
Get the DNSKEY and its RRSIG records from the *DNSKEY RRsets* for the same name server IP. The set of RRSIG records may be empty.
```

To me that would be much better.

Reply:
It is ok. I have done change.